clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -518,17 +519,23 @@ private static int parseAcks(String acksString) {
      * @param maxWaitMs The maximum time in ms for waiting on the metadata
      * @return The cluster containing topic metadata and the amount of time we waited in ms
      */
-    private ClusterAndWaitTime waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
+    private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long maxWaitMs) throws InterruptedException {
         // add topic to metadata topic list if it is not there already and reset expiry
         this.metadata.add(topic);
         Cluster cluster = metadata.fetch();
-        if (cluster.partitionsForTopic(topic) != null)
+        List<PartitionInfo> partitions = cluster.partitionsForTopic(topic);
+        // Return cached metadata if we have it and
+        // if the record's partition is either undefined or within the known partition range
+        if (partitions != null && (partition == null || partition.compareTo(partitions.size()) < 0))
             return new ClusterAndWaitTime(cluster, 0);
 
         long begin = time.milliseconds();
         long remainingWaitMs = maxWaitMs;
         long elapsed = 0;
-        while (cluster.partitionsForTopic(topic) == null) {
+        // Resetting partitions to null here allows us to issue a single metadata update in case the cached metadata is stale.
```

Comment:
Correct. Don't remember why I dismissed do/while when I first considered this part.